signer/xpi/xpi.go

```diff
@@ -143,60 +148,121 @@ func (s *PKCS7Signer) Config() signer.Configuration {
+		return nil, errors.Wrap(err, "xpi: cannot make JAR manifest from XPI")
+	}
+
+	if len(coseSigAlgs) < 1 {
+		pkcs7Manifest = manifest
+	} else {
+		cn, err := opt.CN(s)
+		if err != nil {
+			return nil, err
+		}
+		coseSig, err = s.issueCOSESignature(cn, manifest, coseSigAlgs)
+		if err != nil {
+			return nil, errors.Wrap(err, "xpi: error signing cose message")
+		}
+
+		// add the cose files to the metafiles we'll add to the XPI
+		coseMetaFiles := []Metafile{
+			{"META-INF/cose.manifest", manifest},
+			{"META-INF/cose.sig", coseSig},
+		}
+		metas = append(metas, coseMetaFiles...)
+
+		// add entries for the cose files to the manifest as cose.manifest and cose.sig
+		pkcs7Manifest, err = makePKCS7Manifest(input, metas)
+		if err != nil {
+			return nil, errors.Wrap(err, "xpi: error making PKCS7 manifest")
+		}
+	}
+
+	sigfile, err := makeJARSignature(pkcs7Manifest)
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot make JAR manifest signature from XPI")
+	}
+
+	opt.COSEAlgorithms = []string{}
+	p7sig, err := s.signData(sigfile, opt)
 	if err != nil {
 		return nil, errors.Wrap(err, "xpi: failed to sign XPI")
 	}
-	signedFile, err = repackJAR(input, manifest, sigfile, p7sig)
+
+	metas = append(metas, []Metafile{
+		{"META-INF/manifest.mf", pkcs7Manifest},
+		{"META-INF/mozilla.sf", sigfile},
+		{"META-INF/mozilla.rsa", p7sig},
+	}...)
+
+	signedFile, err = repackJARWithMetafiles(input, metas)
 	if err != nil {
 		return nil, errors.Wrap(err, "xpi: failed to repack XPI")
 	}
 	return signedFile, nil
 }
 
-// SignData takes an input signature file and returns a PKCS7 detached signature
-func (s *PKCS7Signer) SignData(sigfile []byte, options interface{}) (signer.Signature, error) {
-	p7sig, err := s.signData(sigfile, options)
+// SignData takes an input signature file and returns a PKCS7 or COSE detached signature
+func (s *XPISigner) SignData(sigfile []byte, options interface{}) (signer.Signature, error) {
+	opt, err := GetOptions(options)
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot get options")
+	}
+	sigBytes, err := s.signData(sigfile, opt)
 	if err != nil {
 		return nil, err
 	}
 	sig := new(Signature)
-	sig.Data = p7sig
+	sig.Data = sigBytes
 	sig.Finished = true
 	return sig, nil
 }
 
-func (s *PKCS7Signer) signData(sigfile []byte, options interface{}) ([]byte, error) {
-	opt, err := GetOptions(options)
+func (s *XPISigner) signData(sigfile []byte, opt Options) ([]byte, error) {
+	cn, err := opt.CN(s)
 	if err != nil {
-		return nil, errors.Wrap(err, "xpi: cannot get options")
+		return nil, err
 	}
-	cn := opt.ID
-	if s.EndEntityCN != "" {
-		cn = s.EndEntityCN
+	coseSigAlgs, err := opt.Algorithms()
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot make JAR manifest from XPI")
 	}
-	if cn == "" {
-		return nil, errors.New("xpi: missing common name")
+	if len(coseSigAlgs) > 0 {
+		return nil, fmt.Errorf("xpi: cannot use /sign/data for COSE signatures. Use /sign/file instead")
```

Comment:
I don't think this check belongs here. I'd suggest moving it to `SignData` directly, so this function doesn't have to know anything about COSE.

Reply:
yeah this removes the need to set algs to an empty list before calling it from SignFile too